### ISSUE TITLE
Rounding with Integer inputs returns Floats

### ISSUE THREAD

**Issue by jordancluts** [JuliaLang/julia#31522]
`round(1234,sigdigits=2)` returns a Float64 instead of an Int (return value is 1200.0 instead of 1200). When the output type is specified `round (Int,1234,sigdigits=2)` a method error is returned as the sigdigits keyword is not implemented when input is an Int and return type is specified.

**Comment by tkluck** [comment]
> round(1234,sigdigits=2) returns a Float64 instead of an Int (return value is 1200.0 instead of 1200).

This is necessary for type stability: we want `round(1, sigdigits=2)` to be of the same type as `round(12, sigdigits=2)` and as `round(1200, sigdigits=2)`. So `Float64` is really the only option here.

> When the output type is specified round (Int,1234,sigdigits=2) a method error is returned as the sigdigits keyword is not implemented when a return type is specified. Even Float return types are incompatible with sigdigits. round(Float64,1234.5,sigdigits=4) throws the same error.

What made you think that would work? I can't find this signature documented anywhere. (My guess is that you are comparing this e.g. to `zeros(Int, 4,4)` but `round` is not a constructor.) For this use case, just `Int(round(1234, sigdigits=2))` is best.

**Comment by jordancluts** [comment]
I hadn't realized that `round(1,sigdigits=2)` "added" a significant figure to create 1.0. I atleast understand the reasoning behind why one would do it that way. I had assumed that sigdigits on reduced as typically one cannot increase significance. I may disagree with the choice to allow that but given that premise the float return type makes a bit more sense. Perhaps it should be documented a bit better since without trial and error it's fairly difficult to know when one will get an Integer or a Float back from round when an Integer is supplied. (`round(1234)` returns an Int).

> What made you think that would work? I can't find this signature documented anywhere.

I'm sure it's not documented anywhere. I merely got there by knowing that `round(Int,5.678)` works and `round(5.678,sigdigits=2)` works and assumed I could do both. I think however that it's a reasonable and clear syntax and so perhaps implementing the signature to do so would be a useful enhancement.

**Comment by JeffBezanson** [comment]
Agreed; it seems like `round(Float64,1234.5,sigdigits=4)` at least should work.

It also seems reasonable to me for `round(::Integer, sigdigits=n)` to return an integer. I suspect the way this happened was we saw `round` as ` floating-point sort of function, e.g. due to its being in the C math library for floating-point types.

**Comment by tkluck** [comment]
> I had assumed that sigdigits on reduced as typically one cannot increase significance.

Yes, I see your point. One way to think about it is that `Int` has an infinite number of significant digits: Int(3) is equal to 3, not 3 +/- 0.000001.

> I merely got there by knowing that round(Int,5.678) works and round(5.678,sigdigits=2) works and assumed I could do both

My bad, I should have tried those -- your expectation is reasonable.

For my personal preference, I'd argue that `Int(1200)` is not a very good representation of 1,2 * 10^2, precisely because `Int` implies infinite precision. But OTOH, there's no real other way to interpret what the user would want from `round(Int, 1234, sigdigits=2)`. Moreover, there's arguably a similar problem with returning a `Float64` as it also implies more significant digits than 2. So I don't want to be gatekeeping here :)

**Comment by tkluck** [comment]
@jordancluts depending on how badly you want this, you could consider submitting a pull request. You can add the method that you want in `base/floatfuncs.jl`. Feature freeze for v1.2 [is this Thursday](https://discourse.julialang.org/t/julia-1-2-feature-freeze-april-4th/22478) so you could even try to aim for that!
